projects/storefrontlib/src/lib/my-account/update-profile/components/update-profile-form.component.spec.ts

```diff
@@ -0,0 +1,167 @@
+    it('should return true if NOT valid', () => {
+      component.ngOnInit();
+
+      expect(component.isNotValid('firstName')).toEqual(false);
+    });
+
+    it(`should return false if the form was not submitted even though it's invalid`, () => {
+      const invalidUser: User = {
+        ...mockUser,
+        firstName: '',
+      };
+      component.user = invalidUser;
+      component['submitClicked'] = false;
+
+      component.ngOnInit();
+      component.form.get('firstName').markAsDirty();
+
+      expect(component.isNotValid('firstName')).toEqual(false);
+    });
+
+    it('should return false if valid', () => {
+      const invalidUser: User = {
+        ...mockUser,
+        firstName: '',
+      };
+      component.user = invalidUser;
+      component['submitClicked'] = true;
+
+      component.ngOnInit();
+      component.form.get('firstName').markAsDirty();
+
+      expect(component.isNotValid('firstName')).toEqual(true);
+    });
+  });
+
+  describe('onSubmit', () => {
+    it('should call onSubmit() on submit event', () => {
+      spyOn(component, 'onSubmit').and.stub();
+      form.triggerEventHandler('ngSubmit', null);
+      expect(component.onSubmit).toHaveBeenCalled();
+    });
+
+    it('should NOT emit submited event if the form is not valid', () => {
+      spyOn(component.submited, 'emit').and.stub();
+
+      const invalidUser: User = {
+        ...mockUser,
+        firstName: '',
+      };
+      component.user = invalidUser;
+      component.ngOnInit();
+
+      component.onSubmit();
+      expect(component.submited.emit).not.toHaveBeenCalled();
+    });
+
+    it('should emit submited event', () => {
+      spyOn(component.submited, 'emit').and.stub();
+      component.user = mockUser;
+      component.ngOnInit();
+
+      component.onSubmit();
+      expect(component.submited.emit).toHaveBeenCalled();
+    });
+  });
+
+  describe('onSubmit', () => {
+    it('should call onCancel() on click event', () => {
+      spyOn(component, 'onCancel').and.stub();
+      const cancelBtn = el.query(By.css('button[type="button"]'));
+      cancelBtn.triggerEventHandler('click', null);
+      expect(component.onCancel).toHaveBeenCalled();
+    });
+
+    it('should emit cancelled event', () => {
+      spyOn(component.cancelled, 'emit').and.stub();
+      component.onCancel();
+      expect(component.cancelled.emit).toHaveBeenCalled();
+    });
+  });
+
+  describe('when the first name is invalid', () => {
+    const invalidUser: User = {
+      ...mockUser,
+      firstName: '',
+    };
+
+    it('should display an error message', () => {
+      component.user = invalidUser;
+      component.ngOnInit();
+      fixture.detectChanges();
+
+      const error = el.query(By.css('.invalid-feedback'));
+      expect(error).toBeTruthy();
+
+      const message = error.query(By.css('span'));
+      expect(message).toBeTruthy();
+      expect(message.nativeElement.innerText).toEqual(
+        'First name is required.'
```

Comment:
I think the tests would be less brittle if we validated that there _is_ a message for the first name error, without expecting a specific message (because these can evolve as part of properites/label files)

Reply:
fixed